lib/types.nix

```diff
@@ -614,6 +620,102 @@ rec {
       nestedTypes.elemType = elemType;
     };
 
+    attrTag = tags: attrTagWith { inherit tags; };
+
+    attrTagWith = args@{ tags }:
+      let
+        tags =
+          mapAttrs
+            (n: opt:
+              builtins.addErrorContext "while checking that attrTag tag ${lib.strings.escapeNixIdentifier n} is an option with a type${inAttrPosSuffix args.tags n}" (
+                throwIf (opt._type or null != "option")
+                  "In attrTag/attrTagWith, each tag value must be an option, but tag ${lib.strings.escapeNixIdentifier n} ${
+                    if opt?_type then
+                      if opt._type == "option-type"
+                      then "was a bare type, not wrapped in mkOption."
+                      else "was of type ${lib.strings.escapeNixString opt._type}."
+                    else "was not."}"
+                opt // {
+                  declarations = opt.declarations or (
+                    let pos = builtins.unsafeGetAttrPos n args.tags;
+                    in if pos == null then [] else [ pos.file ]
+                  );
+                  declarationPositions = opt.declarationPositions or (
+                    let pos = builtins.unsafeGetAttrPos n args.tags;
+                    in if pos == null then [] else [ pos ]
+                  );
+                }
+              ))
+            args.tags;
```

Comment:
I've shadowed the original `tags` binding from the argument to make sure we use the checked and normalized one below.

Reply:
I generally prefer an about-equally-ugly-but-more-explicit `let _tags = tags; in let tags = f _tags;`, but this is fine too

Reply:
Removing `attrTagWith` forced me to do it like this.

I believe that approach is slightly more expensive because of an extra `Env`, but this isn't hot code, so I'll just stop talking now :)